src/js/view/controls/components/menu/settings-menu.js

```diff
@@ -3,260 +3,458 @@ import { MenuItem } from 'view/controls/components/menu/menu-item';
-                settingsMenu.removeMenu('playbackRates');
+            if (children.playbackRates) {
+                this.removeMenu('playbackRates');
             }
             return;
         }
         const initialSelectionIndex = playbackRates.indexOf(model.get('playbackRate'));
         const menuItemOptions = { tooltipText: localization.playbackRates };
-        changeMenuItems(
+
+        this.setupMenu(
             'playbackRates', 
             playbackRates, 
-            (playbackRate) => api.setPlaybackRate(playbackRate), 
+            (playbackRate) => playerAPI.setPlaybackRate(playbackRate), 
             initialSelectionIndex, 
             menuItemOptions
         );
-    };
-    model.change('playbackRates', (changedModel, playbackRates) => {
-        setPlaybackRatesMenu(playbackRates);
-    }, settingsMenu);
-    model.change('playbackRate', (changedModel, playbackRate) => {
+    }
+
+    onPlaybackRate(model, playbackRate) {
         const rates = model.get('playbackRates');
         let index = -1;
         if (rates) {
             index = rates.indexOf(playbackRate);
         }
-        onMenuItemSelected(settingsMenu.children.playbackRates, index);
-    }, settingsMenu);
+        selectMenuItem(this.children.playbackRates, index);
+    }
 
-    model.on('change:playbackRateControls', () => {
-        setPlaybackRatesMenu(model.get('playbackRates'));
-    });
+    onChangePlaybackRateControls(model) {
+        this.onPlaybackRates(model);
+    }
 
-    // Remove the audio tracks, qualities, and playback rates submenus when casting
-    model.on('change:castActive', (changedModel, active, previousState) => {
+    onChangeCastActive(model, active, previousState) {
         if (active === previousState) {
             return;
         }
 
         if (active) {
-            settingsMenu.removeMenu('audioTracks');
-            settingsMenu.removeMenu('quality');
-            settingsMenu.removeMenu('playbackRates');
+            this.removeMenu('audioTracks');
+            this.removeMenu('quality');
+            this.removeMenu('playbackRates');
         } else {
-            setAudioTracksMenu(model.get('audioTracks'));
-            setLevelsMenu(model.get('levels'));
-            setPlaybackRatesMenu(model.get('playbackRates'));
+            this.audioTracksMenu(model.get('audioTracks'));
+            this.levelsMenu(model.get('levels'));
+            this.playbackRatesMenu(model.get('playbackRates'));
+        }
+    }
+
+    onChangeStreamType() {
+        this.onPlaybackRates(this.model);
+    }
+
+    documentClick(evt) {
+        if (!/jw-(settings|video|nextup-close|sharing-link|share-item)/.test(evt.target.className)) {
+            this.close();
+        }
+    }
+
+    addEventListeners() {
+        const { updateControlbarButtons, model } = this;
+        const handlers = Object.getOwnPropertyNames(Object.getPrototypeOf(this)).filter(name => /^on/.test(name));
```

Comment:
yikes. I'd rather see an object (map) with properties and methods as the key-value pairs. It would be easier to read and iterate through.

This is a lot of work to 'magically' make any `onChange...` method get a property change listener, and the `charAt` `toLowerCase` stuff makes me worry that it would be very easy to miss a listener because of a typo.